tests/e2e/default/targets/target-aggregates.wdio-spec.js

```diff
@@ -109,57 +61,245 @@ describe('Target aggregates', () => {
+          owner: contact._id,
+          user: 'irrelevant',
+        }));
+      }));
+
+    const DOCS_TO_KEEP = [
+      districtHospital1._id,
+      districtHospital2._id,
+      ...docs.map(doc => doc._id),
+      'fixture:user:supervisor',
+      'org.couchdb.user:supervisor',
+      '^target~',
+      [/^form:/],
+    ];
+
+    const getLastMonth = () => {
+      const newDate = new Date();
+      newDate.setDate(1);
+      newDate.setMonth(newDate.getMonth() - 1);
+      return newDate.toLocaleString('default', { month: 'long' });
+    };
+
+    const assertTitle = async (itemTitle, targetTitle) => {
+      expect(itemTitle).to.equal(targetTitle);
+      expect(await (await targetAggregatesPage.targetDetail.title(targetTitle)).isDisplayed()).to.be.true;
+    };
+
+    const assertCounter = async (itemCounter, targetCounter) => {
+      expect(itemCounter).to.equal(targetCounter);
+      expect(await (await targetAggregatesPage.targetDetail.counter()).getText()).to.equal(targetCounter);
+    };
+
+    const assertPlace = async (itemPlace, place) => {
+      expect(itemPlace).to.be.true;
+      expect(await (await targetAggregatesPage.targetDetail.place(place)).isDisplayed()).to.be.true;
+    };
+
+    const assertPeriod = async (itemPeriod, period) => {
+      expect(itemPeriod).to.be.true;
+      expect(await (await targetAggregatesPage.targetDetail.period(period)).isDisplayed()).to.be.true;
+    };
+
+    const assertData = async (contacts, expectedTargets, period, place, assertContactValues = true) => {
+      await targetAggregatesPage.openTargetDetails(expectedTargets[0]);
+      expect(await targetAggregatesPage.aggregateList().length).to.equal(expectedTargets.length);
+      for (const target of expectedTargets) {
+        const targetItem = await targetAggregatesPage.getTargetItem(target, period, place);
+        await targetAggregatesPage.openTargetDetails(target);
+        await assertTitle(targetItem.title, target.title);
+        if (period === 'This month') {
+          await assertCounter(targetItem.counter, target.counter);
+          if (place !== 'onePlace') {
+            await assertPlace(targetItem.place, place);
+          }
+        } else {
+          await assertPeriod(targetItem.period, period);
+          if (place !== 'onePlace') {
+            await assertPlace(targetItem.place, place);
+          }
+        }
+
+        const targetValuesByContact = place === 'District Hospital 2' ?
+          TARGET_VALUES_BY_CONTACT_DH2 :
+          TARGET_VALUES_BY_CONTACT_DH1;
+
+        const  expectedContacts = contacts[0].counter ? contacts : contacts.map(contact => ({
+          _id: contact._id,
+          name: contact.name,
+          counter: targetValuesByContact[contact.name][target.id].counter,
+          progress: targetValuesByContact[contact.name][target.id].progress,
+        }));
+
+        await assertContacts(expectedContacts, target, assertContactValues);
+      }
+    };
```

Comment:
For reusable code, it's better to avoid `if` statements checking by specific things, because it makes more complex and harder to maintain.  

For example, instead of `place === 'District Hospital 2' ?`, better send the target values by contact by parameter.

And, instead of `period === 'This month'`, send a flag as a parameter `isCurrentPeriod`

Rarely nested `If` statements are needed, you can flat the code. 

When a ternary goes in many lines ` contacts[0].counter ? contacts : contacts.map(contact => ({ ....` better use `if` statement and return early.  Going multi lines with ternary isn't considered good practice because of readability 


```
const assertData = async (contacts, expectedTargets, period, isCurrentPeriod, place, targetValuesByContact, assertTargetPlace, assertContactValues = true) => {
      await targetAggregatesPage.openTargetDetails(expectedTargets[0]);
      expect(await targetAggregatesPage.aggregateList().length).to.equal(expectedTargets.length);
     
    for (const target of expectedTargets) {
        const targetItem = await targetAggregatesPage.getTargetItem(target, period, place);
        await targetAggregatesPage.openTargetDetails(target);
        await assertTitle(targetItem.title, target.title);
        
       if (isCurrentPeriod) {
          await assertCounter(targetItem.counter, target.counter);
        } else {
          await assertPeriod(targetItem.period, period);
        }

          if (assertTargetPlace) { // better than  if (place !== 'onePlace') {
            await assertPlace(targetItem.place, place);
          }

        if (contacts[0].counter) {
           await assertContacts(contacts, target, assertContactValues);
           continue;
        }

        const  expectedContacts = contacts.map(contact => ({
          _id: contact._id,
          name: contact.name,
          counter: targetValuesByContact[contact.name][target.id].counter,
          progress: targetValuesByContact[contact.name][target.id].progress,
        }));
        await assertContacts(expectedContacts, target, assertContactValues);
      }
    };
```

Reply:
You are completely right, thank you for the explanation ⭐